### PR TITLE
Introduce citus_calculate_gpid and citus_backend_gpid

### PR DESCRIPTION
**Pull request by onderkalaci**
In some cases, it is useful to see what the GPID a process is going to be assigned by an external process.
In the end, this is a consistent function that generates the same output.

We'll be using this UDF when a backend is blocked o an a DDL and has not been assigned a gpid.

DESCRIPTION: Introduces citus_backend_gpid()

### CODE REVIEW

File: src/backend/distributed/sql/udfs/citus_calculate_gpid/latest.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+CREATE FUNCTION pg_catalog.citus_calculate_gpid(nodeid integer,
+                                                pid integer)
+    RETURNS BIGINT
+    LANGUAGE C STRICT
+    AS 'MODULE_PATHNAME',$$citus_calculate_gpid$$;
+COMMENT ON FUNCTION pg_catalog.citus_calculate_gpid(nodeid integer, pid integer)
+    IS 'activate a node which is in the cluster';
+
+GRANT EXECUTE ON FUNCTION pg_catalog.citus_calculate_gpid(integer, integer) TO PUBLIC;
```
Comment (JelteF) [inline src/backend/distributed/sql/udfs/citus_calculate_gpid/latest.sql:9]: Do we also have a way to get the gpid of the current backend? i.e. the equivalent of `pg_backend_pid()`, so e.g. `citus_backend_gpid()`?

File: src/backend/distributed/sql/udfs/citus_calculate_gpid/latest.sql [hunk 2]
```diff
@@ -0,0 +1,9 @@
+CREATE FUNCTION pg_catalog.citus_calculate_gpid(nodeid integer,
+                                                pid integer)
+    RETURNS BIGINT
+    LANGUAGE C STRICT
+    AS 'MODULE_PATHNAME',$$citus_calculate_gpid$$;
+COMMENT ON FUNCTION pg_catalog.citus_calculate_gpid(nodeid integer, pid integer)
+    IS 'activate a node which is in the cluster';
```
Comment (JelteF) [inline src/backend/distributed/sql/udfs/citus_calculate_gpid/latest.sql:7]: comment is incorrect

File: src/backend/distributed/sql/udfs/citus_calculate_gpid/latest.sql (hunk 1 above)
Comment (onderkalaci) [reply src/backend/distributed/sql/udfs/citus_calculate_gpid/latest.sql:9]: Nope, we intended to have and and forgot. 

I can also get that and use it in the tests

**Comment by halilozanakgul** [comment]
We know it is a calculation but from a user's perspective I would expect something like `citus_get/search/find_gpid`.

File: src/backend/distributed/transaction/backend_data.c [hunk 3]
```diff
@@ -880,10 +883,28 @@ GetGlobalPID(void)
...
- * GenerateGlobalPID generates the global process id for the current backend.
+ * citus_calculate_gpid calculates the gpid for any given process on any
+ * given node.
+ */
+Datum
+citus_calculate_gpid(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	/* prepare data before acquiring spinlock to protect against errors */
```
Comment (halilozanakgul) [inline src/backend/distributed/transaction/backend_data.c:894]: Where is the spinlock?

File: src/test/regress/sql/global_cancel.sql [hunk 4]
```diff
@@ -48,4 +48,6 @@ SELECT pg_terminate_backend(10000000000 * :coordinator_node_id + 0);
 
 RESET client_min_messages;
 
+SELECT citus_backend_gpid() = citus_calculate_gpid(:coordinator_node_id, pg_backend_pid());
```
Comment (halilozanakgul) [inline src/test/regress/sql/global_cancel.sql:51]: Lovely test ❤️

**Review by halilozanakgul — CHANGES_REQUESTED** [review]
Looks good but I'm just concerned about the naming.
We know that it is a calculation but from a user's perspective I would expect something like `citus_get/search/find_gpid`

**Comment by onderkalaci** [comment]
> We know that it is a calculation but from a user's perspective I would expect something like citus_get/search/find_gpid

I think get/search/find do not really reflect what we do. The other alternative I considered was `citus_generate_gpid_for(nodeId,Pid)`. Or, I can change to `citus_calculate_gpid_for(nodeId, pid)`?

File: src/test/regress/sql/global_cancel.sql (hunk 4 above)
Comment (onderkalaci) [reply src/test/regress/sql/global_cancel.sql:54]: look at the other test, it is even nicer:p

**Review by halilozanakgul — APPROVED** [review]

**Pull request merged**